File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/SettingsPage/SettingsFragment.java
package org.wycliffeassociates.translationrecorder.SettingsPage;

import android.app.Activity;
import android.app.FragmentManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;

import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.resync.ResyncLanguageNamesTask;
import org.wycliffeassociates.translationrecorder.TranslationRecorderApp;
import org.wycliffeassociates.translationrecorder.database.ProjectDatabaseHelper;
import org.wycliffeassociates.translationrecorder.utilities.TaskFragment;

import org.wycliffeassociates.translationrecorder.R;

/**
 * Created by leongv on 12/17/2015.
 */
public class SettingsFragment extends PreferenceFragment  implements SharedPreferences.OnSharedPreferenceChangeListener {

    private int FILE_GET_REQUEST_CODE = 45;

    LanguageSelector mParent;
    SharedPreferences mSharedPreferences;
    ProjectDatabaseHelper db;
    private TaskFragment mTaskFragment;
    private String TAG_TASK_FRAGMENT = "tag_task_fragment";

    interface LanguageSelector{
        void sourceLanguageSelected();
    }

    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.preference);
        mSharedPreferences = getPreferenceScreen().getSharedPreferences();
        mParent = (LanguageSelector) getActivity();
        db = ((TranslationRecorderApp)getActivity().getApplication()).getDatabase();
        // Below is the code to clear the SharedPreferences. Use it wisely.
        // mSharedPreferences.edit().clear().commit();

        // Register listener(s)
        mSharedPreferences.registerOnSharedPreferenceChangeListener(this);

        FragmentManager fm = getFragmentManager();
        mTaskFragment = (TaskFragment) fm.findFragmentByTag(TAG_TASK_FRAGMENT);
        if (mTaskFragment == null) {
            mTaskFragment = new TaskFragment();
            fm.beginTransaction().add(mTaskFragment, TAG_TASK_FRAGMENT).commit();
            fm.executePendingTransactions();
        }

        Preference sourceLanguageButton = findPreference(Settings.KEY_PREF_GLOBAL_LANG_SRC);
        sourceLanguageButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                mParent.sourceLanguageSelected();
                return true;
            }
        });

        Preference addTemporaryLanguageButton = findPreference(Settings.KEY_PREF_ADD_LANGUAGE);
        addTemporaryLanguageButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AddTargetLanguageDialog add = new AddTargetLanguageDialog();
                    add.show(getFragmentManager(), "add");
                    return false;
                }
            }
        );

        Preference updateLanguagesButton = findPreference(Settings.KEY_PREF_UPDATE_LANGUAGES);
        updateLanguagesButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Preference updateLanguagesUrlPref = findPreference(Settings.KEY_PREF_UPDATE_LANGUAGES_URL);
                String updateLanguagesUrl = updateLanguagesUrlPref.getSummary().toString();

                mTaskFragment.executeRunnable(
                        new ResyncLanguageNamesTask(1, getActivity(), db, updateLanguagesUrl),
                        "Updating Languages",
                        "Please wait...",
                        true
                );
                return true;
            }
        });

        Preference updateLanguagesFromFileButton = findPreference(Settings.KEY_PREF_UPDATE_LANGUAGES_FROM_FILE);
        updateLanguagesFromFileButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                intent.setType("application/octet-stream");
                startActivityForResult(intent, FILE_GET_REQUEST_CODE);
                return true;
            }
        });

        Preference languagesUrlButton = findPreference(Settings.KEY_PREF_LANGUAGES_URL);
        languagesUrlButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                LanguagesUrlDialog add = new LanguagesUrlDialog();
                add.show(getFragmentManager(), "save");
                return false;
            }
        });

        Preference uploadServerButton = findPreference(Settings.KEY_PREF_UPLOAD_SERVER);
        uploadServerButton.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                UploadServerDialog add = new UploadServerDialog();
                add.show(getFragmentManager(), "save");
                return false;
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == FILE_GET_REQUEST_CODE) {
                Uri uri = resultData.getData();
                mTaskFragment.executeRunnable(
                        new ResyncLanguageNamesTask(1, getActivity(), db, uri),
                        "Updating Languages",
                        "Please wait...",
                        true
                );
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        // Get rid of the extra padding in the settings page body (where it loads this fragment)
        View v = super.onCreateView(inflater, container, savedInstanceState);
        if (v != null) {
            ListView lv = (ListView) v.findViewById(android.R.id.list);
            lv.setPadding(0, 0, 0, 0);
        }
        return v;
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        for (String k : mSharedPreferences.getAll().keySet()) {
            updateSummaryText(mSharedPreferences, k);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
    }

    public void onSharedPreferenceChanged(SharedPreferences mSharedPreferences, String key) {
        updateSummaryText(mSharedPreferences, key);
    }

    private void updateSummariesSetViaActivities(SharedPreferences mSharedPreferences){
        String uristring = mSharedPreferences.getString(Settings.KEY_PREF_GLOBAL_SOURCE_LOC, "");
        Uri dir = Uri.parse(uristring);
        if(dir != null) {
            uristring = dir.getLastPathSegment();
            //This removes "primary:", though maybe this is helpful in identifying between sd card and internal storage.
            //uristring = uristring.substring(uristring.indexOf(":")+1, uristring.length());
            findPreference(Settings.KEY_PREF_GLOBAL_SOURCE_LOC).setSummary(uristring);
        } else {
            findPreference(Settings.KEY_PREF_GLOBAL_SOURCE_LOC).setSummary(mSharedPreferences.getString(Settings.KEY_PREF_GLOBAL_SOURCE_LOC, ""));
        }
    }

    public void updateSummaryText(SharedPreferences mSharedPreferences, String key) {
        try {
            updateSummariesSetViaActivities(mSharedPreferences);
            String text  = mSharedPreferences.getString(key, "");
            if(findPreference(key) != null) {
                findPreference(key).setSummary(text);
            }
        } catch (ClassCastException err) {
            System.out.println("IGNORING SUMMARY UPDATE FOR " + key);
        }
    }
}

File: translationRecorder/app/src/main/java/org/wycliffeassociates/translationrecorder/SettingsPage/SettingsFragment.kt
package org.wycliffeassociates.translationrecorder.SettingsPage

import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import dagger.hilt.android.AndroidEntryPoint
import org.wycliffeassociates.translationrecorder.ProjectManager.activities.ActivityProjectManager.Companion.DATABASE_RESYNC_TASK
import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.CreateBackupTask
import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.MigrateAppFolderTask
import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.RestoreBackupTask
import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.resync.ProjectListResyncTask
import org.wycliffeassociates.translationrecorder.ProjectManager.tasks.resync.ResyncLanguageNamesTask
import org.wycliffeassociates.translationrecorder.R
import org.wycliffeassociates.translationrecorder.SettingsPage.BackupRestoreDialog.BackupRestoreDialogListener
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.BACKUP_TASK_TAG
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_ADD_LANGUAGE
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_BACKUP_RESTORE
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_GLOBAL_LANG_SRC
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_GLOBAL_SOURCE_LOC
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_LANGUAGES_URL
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_MIGRATE_OLD_APP
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_UPDATE_LANGUAGES
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_UPDATE_LANGUAGES_FROM_FILE
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_UPDATE_LANGUAGES_URL
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.KEY_PREF_UPLOAD_SERVER
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.MIGRATE_TASK_TAG
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.RESTORE_TASK_TAG
import org.wycliffeassociates.translationrecorder.SettingsPage.SettingsActivity.Companion.RESYNC_LANGUAGE_NAMES_TASK_TAG
import org.wycliffeassociates.translationrecorder.database.IProjectDatabaseHelper
import org.wycliffeassociates.translationrecorder.persistance.AssetsProvider
import org.wycliffeassociates.translationrecorder.persistance.IDirectoryProvider
import org.wycliffeassociates.translationrecorder.persistance.IPreferenceRepository
import org.wycliffeassociates.translationrecorder.persistance.getDefaultPref
import org.wycliffeassociates.translationrecorder.project.ChunkPluginLoader
import org.wycliffeassociates.translationrecorder.usecases.CreateBackup
import org.wycliffeassociates.translationrecorder.usecases.MigrateOldApp
import org.wycliffeassociates.translationrecorder.usecases.RestoreBackup
import org.wycliffeassociates.translationrecorder.utilities.TaskFragment
import java.io.File
import javax.inject.Inject

/**
 * Created by leongv on 12/17/2015.
 */
@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat(),
    OnSharedPreferenceChangeListener, BackupRestoreDialogListener {

    @Inject lateinit var db: IProjectDatabaseHelper
    @Inject lateinit var prefs: IPreferenceRepository
    @Inject lateinit var directoryProvider: IDirectoryProvider
    @Inject lateinit var assetsProvider: AssetsProvider
    @Inject lateinit var migrateOldApp: MigrateOldApp
    @Inject lateinit var createBackup: CreateBackup
    @Inject lateinit var restoreBackup: RestoreBackup

    private var parent: LanguageSelector? = null
    private lateinit var taskFragment: TaskFragment

    private lateinit var openDirectory: ActivityResultLauncher<Uri?>
    private lateinit var openLanguagesFile: ActivityResultLauncher<String>

    interface LanguageSelector {
        fun sourceLanguageSelected()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        taskFragment = (parentFragmentManager.findFragmentByTag(TAG_TASK_FRAGMENT) as? TaskFragment) ?: run {
            val fragment = TaskFragment()
            parentFragmentManager.beginTransaction().add(fragment, TAG_TASK_FRAGMENT).commit()
            fragment
        }

        val handler = Handler(Looper.getMainLooper())
        activity?.intent?.let { handler.post { handleIntent(it) } }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        view.fitsSystemWindows = true

        openDirectory = registerForActivityResult(ActivityResultContracts.OpenDocumentTree()) {
            it?.let(::migrate)
        }

        openLanguagesFile = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
            uri?.let {
                taskFragment.executeRunnable(
                    ResyncLanguageNamesTask(
                        RESYNC_LANGUAGE_NAMES_TASK_TAG,
                        requireContext(),
                        db,
                        assetsProvider,
                        requireContext(),
                        it
                    ),
                    getString(R.string.updating_languages),
                    getString(R.string.please_wait),
                    true
                )
            }
        }

        return view
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        parent = activity as LanguageSelector

        preferenceManager.createPreferenceScreen(requireContext())
        addPreferencesFromResource(R.xml.preference)

        updateSummaryText(KEY_PREF_GLOBAL_LANG_SRC)
        updateSummaryText(KEY_PREF_LANGUAGES_URL)
        updateSummaryText(KEY_PREF_UPLOAD_SERVER)

        val globalLanguagePref: Preference? = findPreference(KEY_PREF_GLOBAL_LANG_SRC)
        globalLanguagePref?.setOnPreferenceClickListener {
            parent?.sourceLanguageSelected()
            true
        }

        val addLanguagePref: Preference? = findPreference(KEY_PREF_ADD_LANGUAGE)
        addLanguagePref?.setOnPreferenceClickListener {
            val add = AddTargetLanguageDialog()
            add.show(parentFragmentManager, "add")
            true
        }

        val updateLanguagesUrlPref: Preference? = findPreference(KEY_PREF_UPDATE_LANGUAGES)
        updateLanguagesUrlPref?.setOnPreferenceClickListener {
            val updateLanguageUrlPref: Preference? = findPreference(KEY_PREF_UPDATE_LANGUAGES_URL)
            val updateLanguagesUrl = updateLanguageUrlPref?.summary?.toString()
            updateLanguagesUrl?.let { url ->
                taskFragment.executeRunnable(
                    ResyncLanguageNamesTask(
                        RESYNC_LANGUAGE_NAMES_TASK_TAG,
                        requireContext(),
                        db,
                        assetsProvider,
                        requireContext(),
                        url
                    ),
                    getString(R.string.updating_languages),
                    getString(R.string.please_wait),
                    true
                )
            }
            true
        }

        val updateLanguagesFromFilePref: Preference? = findPreference(KEY_PREF_UPDATE_LANGUAGES_FROM_FILE)
        updateLanguagesFromFilePref?.setOnPreferenceClickListener {
            openLanguagesFile.launch("*/*")
            true
        }

        val languagesUrlPref: Preference? = findPreference(KEY_PREF_LANGUAGES_URL)
        languagesUrlPref?.setOnPreferenceClickListener {
            val add = LanguagesUrlDialog()
            add.show(parentFragmentManager, "save")
            true
        }

        val uploadServerPref: Preference? = findPreference(KEY_PREF_UPLOAD_SERVER)
        uploadServerPref?.setOnPreferenceClickListener {
            val add = UploadServerDialog()
            add.show(parentFragmentManager, "save")
            true
        }

        val migratePref: Preference? = findPreference(KEY_PREF_MIGRATE_OLD_APP)
        migratePref?.setOnPreferenceClickListener {
            openDirectory.launch(null)
            true
        }

        val backupRestorePref: Preference? = findPreference(KEY_PREF_BACKUP_RESTORE)
        backupRestorePref?.setOnPreferenceClickListener {
            val dialog = BackupRestoreDialog()
            dialog.setListener(this)
            dialog.show(parentFragmentManager, "backup")
            true
        }
    }

    private fun migrate(uri: Uri) {
        taskFragment.executeRunnable(
            MigrateAppFolderTask(MIGRATE_TASK_TAG, uri, migrateOldApp),
            getString(R.string.migrating_old_app),
            getString(R.string.please_wait),
            true
        )
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        key?.let { updateSummaryText(it) }
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences?.let { preferences ->
            preferences.registerOnSharedPreferenceChangeListener(this)
            preferences.all.keys.forEach {
                updateSummaryText(it)
            }
        }
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen
            .sharedPreferences
            ?.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onCreateBackup(zipFileUri: Uri) {
        taskFragment.executeRunnable(
            CreateBackupTask(BACKUP_TASK_TAG, zipFileUri, createBackup),
            getString(R.string.creating_backup),
            getString(R.string.please_wait),
            true
        )
    }

    override fun onRestoreBackup(zipFileUri: Uri) {
        taskFragment.executeRunnable(
            RestoreBackupTask(RESTORE_TASK_TAG, zipFileUri, restoreBackup),
            getString(R.string.restoring_backup),
            getString(R.string.please_wait),
            true
        )
    }

    private fun updateSummaryText(key: String) {
        try {
            var summary = prefs.getDefaultPref(key, "")
            val pref = findPreference(key) as? Preference

            when (key) {
                KEY_PREF_GLOBAL_SOURCE_LOC -> {
                    val file = File(summary)
                    if (file.exists()) {
                        summary = file.name
                    }
                }
                KEY_PREF_LANGUAGES_URL -> {
                    if (summary.isEmpty()) {
                        summary = getString(R.string.pref_languages_url)
                    }
                }
                KEY_PREF_UPLOAD_SERVER -> {
                    if (summary.isEmpty()) {
                        summary = getString(R.string.pref_upload_server)
                    }
                }
            }
            pref?.summary = summary
        } catch (e: ClassCastException) {
            println("IGNORING SUMMARY UPDATE FOR $key")
        }
    }

    fun resyncProjectList() {
        val task = ProjectListResyncTask(
            DATABASE_RESYNC_TASK,
            parentFragmentManager,
            db,
            directoryProvider,
            ChunkPluginLoader(directoryProvider, assetsProvider),
            true
        )
        taskFragment.executeRunnable(
            task,
            getString(R.string.resyncing_database),
            getString(R.string.please_wait),
            true
        )
    }

    private fun handleIntent(intent: Intent?) {
        when (intent?.action) {
            Intent.ACTION_SEND -> {
                if (intent.type?.startsWith("application/zip") == true) {
                    val fileUri = intent.getParcelableExtra<Uri>(Intent.EXTRA_STREAM)
                    fileUri?.let { uri: Uri ->
                        onRestoreBackup(uri)
                    }
                }
            }
        }
    }

    private companion object {
        const val TAG_TASK_FRAGMENT = "tag_task_fragment"
    }
}